Add handler rotation and expiry

rynowak opened the pull request:
This adds a simple approach to handler rotation and handler expiry that
should have virtually no impact on how users consume HttpClient.

The idea is that handlers sit in an 'active' pool until they expire
(with a configuable fixed lifetime). While they are in the 'active' pool
they can be shared amongst as many HttpClient instances as we need to
create.

When a handler that supports expiry is created, a timer will be queued
to atomically remove it from the pool. A new one will then be created to
take its place.

----

Ultimately we need to dispose the handler that we removed from the
active pool, but it's not safe to do that right away because there are
probably existing leases. Ultimately this is a task for the GC.

When we create the handler, the one that we give to HttpClients is an
instance of a type we control, that is safe to GC without the need for
disposal.

Once we move the handler to the 'expiry' pool, we can downgrade our
reference to the tracking handler to a weak reference. Once we see that
its GCed, it's safe to dispose the inner handler.

----

We ultimately process all of these disposal jobs on a fixed timer that
processes all of the 'expired' items.

Since all of the data flow is unidirectional there isn't much threading
heavy-lifting.

rynowak commented:
Not included here:
- tests
- updates to the builder API

rynowak commented:
/cc @anurse @glennc PTAL

rynowak commented:
> When we create the handler, the one that we give to HttpClients is an instance of a type we control, that is safe to GC without the need for disposal.

I also considered the approach where we keep a weakreference for each HttpClient we create, but I didn't like how the math works out in high throughput scenarios. Each 'expired' item could have thousands or millions or weakreferences associated with it. 

This approach is a little sneaky, but allows the GC to do all of the work for us. It's basically like using a finalizer in the end, but without the gotchas of finalizers. It's more deterministic, because we can control how often we clean up.

rynowak commented on src/Microsoft.Extensions.Http/ActiveHandlerTrackingEntry.cs:
```diff
@@ -0,0 +1,72 @@
+                return;
+            }
+
+            StartExpiryTimerSlow(callback);
```
This is basically an implementation of `LazyInitializer.EnsureInitialized`.

rynowak commented on src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs:
```diff
@@ -87,16 +144,154 @@ internal HttpMessageHandler CreateHandler(string name)
+            {
+                if (_cleanupTimer == null)
+                {
+                    _cleanupTimer = new Timer(_cleanupCallback, null, DefaultCleanupInterval, Timeout.InfiniteTimeSpan);
```
note: I prefer this pattern with timers - basically setting an alarm that will go off once at some point in the future. Since there's nothing recurring, you don't have any state to dispose. You just clean up each timer as it goes off.

The BCL already optimizes for using lots of short-lived timers.

anurse replied:
Good to know. Might want to do something similar with timeouts in SignalR :)

rynowak commented on src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs:
```diff
@@ -87,16 +144,154 @@ internal HttpMessageHandler CreateHandler(string name)
+            Log.CleanupCycleStart(_logger, initialCount);
+
+            var disposedCount = 0;
+            for (var i = 0; i < initialCount; i++)
```
There's actually a subtle issue here, we need to prevent cleanup cycles from running concurrently if the time the cycle takes is longer than the period. 

We really don't intend for multiple threads to do cleanup at once. I could make it supported, but it seems better to prevent it.

Will fix in the next iteration.

NTaylorMullen commented on src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs:
```diff
@@ -87,16 +144,154 @@ internal HttpMessageHandler CreateHandler(string name)
+            }
+        }
+
+        private static class Log
```
I assume you'll pull this out in the non-design version.

rynowak replied:
No this is the pattern we're using. It's much cleaner than a single class with infinity methods

anurse replied:
@BrennanConroy we should do this in SignalR.

BrennanConroy replied:
I know, you made an issue for it, after preview1

rynowak commented on src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs:
```diff
@@ -87,16 +144,154 @@ internal HttpMessageHandler CreateHandler(string name)
+        internal void ExpiryTimer_Tick(object state)
+        {
+            var active = (ActiveHandlerTrackingEntry)state;
+            active.DisposeTimer();
```
How would the layering of this work?

NTaylorMullen replied:
Can have a wrapping delegate in your` ActiveHandlerTrackingEntry` which is responsible for calling into timercallback code.

rynowak replied:
I chose to address this for now. It results in an additional delegate per handler entry. We might undo this later.

anurse commented on src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs:
```diff
@@ -87,16 +150,195 @@ internal HttpMessageHandler CreateHandler(string name)
+            // 
+            // Otherwise it would be possible that we start the timer here, immediately expire it (very short
+            // timer) and then dispose it without ever creating a client. That would be bad. It's unlikely
+            // this would happen, just we want to be sure.
```
nit: ~just~ but

anurse commented on src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs:
```diff
@@ -87,16 +150,195 @@ internal HttpMessageHandler CreateHandler(string name)
+            // our entry in the collection, then this is a bug.
+            var removed = _activeHandlers.TryRemove(active.Name, out var found);
+            Debug.Assert(removed);
+            Debug.Assert(object.ReferenceEquals(active, found.Value));
```
nit: Messages in the asserts?

anurse commented on src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs:
```diff
@@ -87,16 +150,195 @@ internal HttpMessageHandler CreateHandler(string name)
+            entry.StartExpiryTimer(ExpiryTimer_Tick);
+        }
+
+        // Internal so it can be overridden in tests
```
`private protected`? #usethemnewfeatures

anurse replied:
FYI, it does work with `InternalsVisibleTo`, to my knowledge.

rynowak replied:
We're not using 7.2 here yet - I'd rather wait until that's turned on globally.

anurse commented on src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs:
```diff
@@ -87,16 +150,195 @@ internal HttpMessageHandler CreateHandler(string name)
+                {
+                    // Since we're the only one removing from _expired, TryDequeue must always succeed.
+                    _expiredHandlers.TryDequeue(out var entry);
+                    Debug.Assert(entry != null);
```
nit: message on assert

anurse commented on src/Microsoft.Extensions.Http/Logging/EventIds.cs:
```diff
@@ -23,5 +23,10 @@ internal static class EventIds
+        public static readonly EventId FactoryCleanupCycleStart = new EventId(100, "CleanupCycleStart");
+        public static readonly EventId FactoryCleanupCycleEnd = new EventId(101, "CleanupCycleEnd");
+        public static readonly EventId FactoryCleanupItemFailed = new EventId(102, "CleanupItemFailed");
+        public static readonly EventId FactoryHandlerExpired = new EventId(103, "HandlerExpired");
```
Hrm, didn't notice this fully in previous reviews. Why is this not part of the `Log` nested class? Aren't we just substituting one large comment-delimited class for another? Also, putting them in `Log` means the names can use `nameof`.

rynowak replied:
yeah I think you're right about this. I'm not going to use nameof though. I want these to be hardcoded constants because we're not allowed to change them.

anurse replied:
> I want these to be hardcoded constants because we're not allowed to change them.

Good point.

rynowak commented:
@anurse @NTaylorMullen updated

NTaylorMullen commented on src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs:
```diff
@@ -4,27 +4,66 @@
+        // 10 distinct named clients * expiry time >= 1s = approximate cleanup queue of 300 items
+        //
+        // This seems frequent enough. We also rely on GC occurring to actually trigger disposal.
+        private readonly TimeSpan DefaultCleanupInterval = TimeSpan.FromSeconds(10);
```
Need to update comment.

NTaylorMullen commented on src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs:
```diff
@@ -4,27 +4,66 @@
     internal class DefaultHttpClientFactory : IHttpClientFactory
     {
+        // establishing a minimum lifetime helps us avoid some possible destructive cases.
+        private readonly TimeSpan MinimumHandlerLifetime = TimeSpan.FromSeconds(1);
```
You don't use this anymore.

NTaylorMullen commented on test/Microsoft.Extensions.Http.Test/DefaultHttpClientFactoryTest.cs:
```diff
@@ -203,5 +219,279 @@ public void Factory_CreateClient_FiltersCanDecorateBuilder()
+        }
+
+        [Fact]
+        public async Task Factory_CreateClient_WithExpiry_CanHandlerCanBeReusedBeforeExpiry()
```
`Factory_CreateClient_WithExpiry_HandlerCanBeReusedBeforeExpiry` (just changing `CanHandler` => `Handler`)

NTaylorMullen commented on test/Microsoft.Extensions.Http.Test/DefaultHttpClientFactoryTest.cs:
```diff
@@ -203,5 +219,279 @@ public void Factory_CreateClient_FiltersCanDecorateBuilder()
+        }
+
+        [Fact]
+        public async Task Factory_CleanupCycle_DoesDisposeLiveHandler()
```
`DoesDispose` => `Disposes`

NTaylorMullen commented on src/Microsoft.Extensions.Http/Logging/EventIds.cs:
```diff
@@ -8,20 +8,15 @@ namespace Microsoft.Extensions.Http.Logging
     internal static class EventIds
```
This is internal and all event ids are nested now. Suggest removing this entirely.

rynowak replied:
Oh oops, this is supposed to be gone

rynowak commented on src/Microsoft.Extensions.Http/DependencyInjection/HttpClientBuilderExtensions.cs:
```diff
@@ -418,5 +419,45 @@ public static IHttpClientBuilder AddTypedClient<TClient>(this IHttpClientBuilder
+                throw new ArgumentNullException(nameof(builder));
+            }
+
+            if (handlerLifetime < TimeSpan.FromSeconds(1))
```
I don't super care about parameterizing the resource, That's a very complex thing to solve. I'll define this in a single place and add a comment to update the resource.

rynowak commented:
> Seems like an EventSource and/or EventCounters might be nice for this, to track active handlers and ensure that they're actually getting rotated properly.

Currently planned for preview-2 https://github.com/aspnet/HttpClientFactory/issues/26 


@anurse - any other feedback? I think I've addressed all of the comments so far
